dirty_cat/datetime_encoder.py

```diff
@@ -0,0 +1,171 @@
+from sklearn.base import BaseEstimator, TransformerMixin
+import numpy as np
+import pandas as pd
+from dirty_cat.utils import check_input
+from pandas.tseries.holiday import USFederalHolidayCalendar as calendar
+from typing import Union, Optional, List
+
+
+
+class DatetimeEncoder(TransformerMixin, BaseEstimator):
+    """
+    This encoder transforms each datetime column into several numeric columns corresponding to temporal features,
+    e.g year, month, day...
+    Constant extracted features are dropped ; for instance, if the year is always the same in a feature, the extracted "year" column won't be added.
+    If the dates are timezone aware, all the features extracted will correspond to the provided timezone.
+
+    Parameters
+    ----------
+    extract_until : {"year", "month", "day", "hour", "minute", "second",
+    "millisecond", "microsecond", "nanosecond"}, default="hour"
+        Extract up to this granularity, and gather the rest into the "other" feature.
+        For instance, if you specify "day", only "year", "month", "day" and "other" features will be created.
+        The "other" feature will be a numerical value expressed in the "extract_until" unit.
+    add_day_of_the_week: bool, default=False
+        Add day of the week feature (if day is extracted). This is a numerical feature from 0 to 6.
+    add_holidays : bool, default=False
+        Whether to add a numerical variable encoding if the day of the date is a holiday (1 for holiday,
+        0 for non-holiday).
+        Uses pandas calendar, which for now only contains US holidays.
+
+    Attributes
+    ----------
+    n_features_out_: int
+        Number of features of the transformed data.
+    features_per_column: Dict[int, List[str]]
+        A dictionary mapping the index of the original columns
+        to the list of features extracted for each column.
+    """
+    def __init__(self,
+                 extract_until="hour",
+                 add_day_of_the_week=False,
+                 add_holidays=False):
+        self.to_extract = ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]
+        self.extract_until = extract_until
+        self.add_day_of_the_week = add_day_of_the_week
+        self.add_holidays = add_holidays
+        # Some functions need aliases
+        self.word_to_alias = {"year": "Y", "month": "M", "day": "D", "hour": "H", "minute": "min", "second": "S",
+                              "millisecond": "ms", "microsecond": "us", "nanosecond": "N"}
```

Comment:
The convention in scikit-learn is that the init should do nothing other than set parameters handed in as inputs. One reason is to have a very well understood statefulness of the objects.

These can be set at the beginning of fit.